src/lib_io.c

```diff
@@ -126,30 +126,46 @@ static int io_file_close(lua_State *L, IOFileUD *iof)
 static int io_file_readnum(lua_State *L, FILE *fp)
 {
   lua_Number d;
-  if (fscanf(fp, LUA_NUMBER_SCAN, &d) == 1) {
-    if (LJ_DUALNUM) {
-      int32_t i = lj_num2int(d);
-      if (d == (lua_Number)i && !tvismzero((cTValue *)&d)) {
-	setintV(L->top++, i);
-	return 1;
-      }
+  int rc;
+  for (;;) {
+    rc = fscanf(fp, LUA_NUMBER_SCAN, &d);
+    if (rc == 1) break;
+    if (ferror(fp)) {
+      if (errno == EINTR) { clearerr(fp); continue; }
+      return 0;
+    } else {
+      setnilV(L->top++);
```

Comment:
Maybe it is no need to set the `L->top` to `nil`, as the `L->top` will be set to `nil` later if the return value is zero.
https://github.com/openresty/luajit2/blob/v2.1-agentzh/src/lib_io.c#L230

Remember to increase `L->top` by one once the return value is zero. Another option is to remove the `if (!ok)` check and set `L->top` to `nil` immediately in all branches.

Reply:
Good catch - I try to minimize the impact of this patch on the upstream source as much as possible. The upstream patch (see https://github.com/LuaJIT/LuaJIT/pull/376) has other improvements to the IO module, but this patch (against the OpenResty fork) focuses solely on EINTR errors. I do not think it is its role to fix other types of behavior if it means the patch size will increase more than needed.

I'll definitely add this to the upstream patch though!

Reply:
(That said, this is my vision of the patch but I am pretty flexible about it, if it is deemed best :) )